<h1 align='center'>Math for Data Science Notes<h1>

<h2>Table of Contents</h2>
<div class='alert alert-block alert-info' style='margin-top: 20px'>
    <ul>
        <li><a href='#rule_of_product'>Rule of Product</a></li>
        <li><a href='#segmentation'>Segmentation</a></li>
        <li><a href='#splitting_datasets'>Splitting Datasets</a></li>
    </ul>

<h3><a id='rule_of_product'>Rule of Product</a></h3>

Q1. How many 5-digit numbers are there that have digits 1, 2, 3, 4 and 5 (each of them exactly once)?)

In [22]:
# manual calculation
x = 0
for i in range(100000):
    s =('0000'+str(i))[-5:] 
    if s.count('1')==1 and s.count('2')==1 and s.count('3')==1 and s.count('4')==1 and s.count('5')==1:
        x+=1

# rule of product
y = 5*4*3*2*1
print('Manual calculation: {}, RoP: {}'.format(x,y))

Manual calculation: 120, RoP: 120


Ans: We have to pick which of five digits we place on the first position, which of the remaining four digits we place on the second position and so on. We are dealing with permutations of numbers 1, 2, 3, 4 and 5 here and the answer is 5*4*3*2*1=120.

Q2. How many integer numbers between 0 and 9999 are there that have exactly one digit 7 and exactly one digit 2?

In [23]:
# manual calculation
x = 0
for i in range(10000):
    s =('0000'+str(i))[-4:] 
    if s.count('2')==1 and s.count('7')==1 :
        x+=1
x

# rule of product
y = 4*3*8*8
print('Manual calculation: {}, RoP: {}'.format(x,y))

Manual calculation: 768, RoP: 768


Ans: There are 4 ways to pick the position for digit 7. Then there are 3 positions remaining for digit 2. And then we can pick each of the remaining digits in 8 possible ways (digits 7 and 2 are forbidden). Thus by the rule of product there are 4x3x8x8=768 possible numbers.

<h3><a id="segmentation">Segmentation</a></h3>

Q1. What is the number of segments in the picture below? (Each segment joins two circles.)


<img src='img/polygon10.png' width="200"/>

Manual calculation will be the sum of x-1 possible segments, i.e:

<code>(10-1) + (9-1) + (8-1) + ... (1-1)</code>

In [6]:
x=0
for i in range(1,11):
    x += i-1
print('x = {}, iteration = {}'.format(x,10))

x = 45, iteration = 10


There is 10 points, thus, by 10 chooses 2 calculation; $\binom{10}{2} = \frac{10x9}{2} = 45$

<h3><a id='splitting_datasets'>Splitting Datasets</a></h3>

Question 1: Suppose we have a dataset of size 12 and we want to construct a subset of data of size 6. How many ways do we have to do it?

Ans: It is just $\binom{12}{6}$

In [20]:
#12 chooses 6
x = 1
y = 1
for i, j in zip(range(12,6,-1),range(6,0,-1)):
    x *= i
    y *= j
x/y


924.0

Question 2: Suppose we have a dataset of size 12 and we want to split it into two subsets of data of size 6 (and it does not matter, which of the subsets is the 
first and which is the second). How many ways do we have to do it?

Ans: There are $\binom{12}{6}=924$ ways of selecting the first subset. When the first subset is selected, the second subset is uniquely determined. But this way, each pair of subsets (A,B) is counted twice: when A is selected as the first subset and when B is selected as the first subset. Thus, the resulting count should be divided by two. This finally gives $\binom{12}{6}/2=462$.

In [21]:
#12 chooses 6 by 2
x = 1
y = 1
for i, j in zip(range(12,6,-1),range(6,0,-1)):
    x *= i
    y *= j
x/y/2


462.0